### DISCUSSION TITLE
`class` vs `typename`

### DISCUSSION THREAD

**Discussion by AlexGuteniev** [microsoft/STL#2288] [Q&A] [answered]
Any reason to prefer `class` over `typename` besides historical?

Sure I mean the context where they are equivalent.

In my code I take advantage of both keyword available: `typename` is something more generic, like a type for a container or smart pointer, that may have few requirements on copyability, etc, whereas `class` is something more specific, like, for example `Clock` or some traits type.

**Comment by StephanTLavavej** [comment, marked as answer]
It's both historical, and reflects the Standard's editorial convention. We don't feel super strongly about it - I personally prefer `typename` (which is why it's common in the tests, which have been less strict about following the same conventions as product code), but I don't think it's worth converting the whole codebase over (unlike clang-format, braces everywhere, etc.).

Using `class` versus `typename` to express a connotation of "only class types" versus "literally any type" could be reasonable for certain codebases that are very strict about it, but because that isn't a widespread convention, I personally don't view the keywords as carrying that connotation (whereas `struct` versus `class` carries stronger connotations of "no secrets" versus "has secrets", although not perfectly).
